### DISCUSSION TITLE
System Prompt Design: Who is user ?

### DISCUSSION THREAD

**Discussion by FilterJoe** [langroid/langroid#555] [General]
One thing I did not understand after reading the architecture blog post:

The system prompts often use the word "user" or "me" when the actual human user isn't involved. For example, in chat-search-assistant, the assistant system prompt is:

You are a resourceful assistant, able to think step by step to answer
        complex questions from the user. You must break down complex questions into
        simpler questions that can be answered by a web search. You must ask me 
        (the user) each question ONE BY ONE, and I will do a web search and send you
        a brief answer. Once you have enough information to answer my original
        (complex) question, you MUST say {DONE} and present the answer to me.

It says "you must ask me (the user) . . . 

Is every agent in a subtask referred to as a user?

Is there something conceptually incorrect if you were to instead have: "you must ask me (the search assistant agent) . . .

What concept(s) do I need to understand to get this right as I design my own system prompts for new types of agents?

**Comment by pchalasani** [comment]
Good question.

> Is every agent in a subtask referred to as a user?

Not exactly. The reason I use "me" is that from the pov of the LLM, it only interacts with the user, so the instructions to the specific agent can say things like "ask me (the user)" or alternatively "ask  me (the search agent)", etc. As far as this agent is concerned, it doesn't matter whether the inputs (or results) come from another agent, or a human user -- they would all have "user" role in its message history. So this type of instruction partly reflects that, and also avoids confusing the LLM. But your suggestion is fine too.
